### ISSUE TITLE
Error while retrieving App Check token: FirebaseError: AppCheck: Requests throttled due to 403 error. Attempts allowed again after 23h:59m:13s (appCheck/throttled)Title for the bug

### ISSUE THREAD

**Issue by mpullerits** [firebase/firebase-js-sdk#7371]
### Operating System

All OS

### Browser Version

All Browser, latest Chrome, latest FF, etc.

### Firebase SDK Version

9.22.2

### Firebase SDK Product:

AppCheck

### Describe your project's tooling

Standard Create-React-App project. Nothing out of the ordinary. 

### Describe the problem

We regularly encounter the error:
Error while retrieving App Check token: FirebaseError: AppCheck: Requests throttled due to 403 error. Attempts allowed again after 23h:59m:13s (appCheck/throttled)

### Steps and code to reproduce issue

We very frequently encounter this error, as well, our customers very frequently encounter it, as per LogRocket recordings.

@firebase/auth:
Auth (9.22.2): Error while retrieving App Check token: FirebaseError: AppCheck: Requests throttled due to 403 error. Attempts allowed again after 23h:59m:13s (appCheck/throttled).

**Comment by mpullerits** [comment]
It is a very random problem. Sometimes it will work totally fine. Then, it will fail. One very odd thing I noticed is in Chrome, in normal browser, often it will work, but then on the same PC, if I open the same URL on a Private Chrome tab, then I will see the appcheck/throttled error!

It is very strange and inconsistent.

I experience the problem reliably when testing from my mobile phone when connected to the wifi. However, when I disconnected from wifi and used cellular data, immediately it started to work totally fine.

Something very strange going on here.

In another LogRocket where the problem also occurred, I see this. Not sure if's helpful to you:

@firebase/app-check:
Failed to read token from IndexedDB. Error: FirebaseError: AppCheck: Error thrown when opening storage. Original error: A mutation operation was attempted on a database that did not allow mutations.. (appCheck/storage-open).

**Comment by cristian-fysight** [comment]
I've been facing this same issue since yesterday.

I disabled app check on my entire app, even in cloud functions, and the error stopped, but right now some users are reporting the issue, and I'm getting some logs about that.

This is the error:
`{
arguments: [
[2023-06-19T22:59:45.354Z]  @firebase/app-check:, 
AppCheck: Requests throttled due to 403 error. Attempts allowed again after 01d:00m:00s (appCheck/throttled).
], 
logger: console
}`

**Comment by dwyfrequency** [comment]
Thank you for logging this issue. Would any of you be able to create a minimal repo to replicate this? It would greatly expedite our ability to see your issue and diagnose any potential pitfalls.

**Comment by laurentpayot** [comment]
Isn’t it the same as https://github.com/firebase/firebase-js-sdk/issues/6708?

**Comment by mpullerits** [comment]
No, because the problem I encounter is not related to sleep.  The problem will occur on the first page load.

**Comment by mpullerits** [comment]
@dwyfrequency Then, sometimes, it will work totally fine in the Chrome private tab!! It is very strange and very inconsistent.

Also, did you see my remark about testing on my mobile - it would fail reliably consistently on my mobile on the Wifi (same network as my PC), however when disabling the Wifi and using cellular data, immediately it will work! 

There must be some kind of IP throttling/hashing on the appcheck side that we're not aware of.

**Comment by hsubox76** [comment]
Sorry about that. Some background on how App Check works on web - I'll try to be verbose here so I can reuse this explanation for other App Check issues.

## ReCAPTCHA and 403s
The default attestation provider used is ReCAPTCHA (V3 or Enterprise) which I assume you are using. The first thing that happens when Firebase tries to get an App Check token is it uses ReCAPTCHA (with the site key you provided) to get a ReCAPTCHA token. If your environment fails the ReCAPTCHA check for any reason, you still get a token, but it's encoded that it failed. It's not designed to be decodable on the client side for obvious reasons (ReCAPTCHA is supposed to catch bad actors and you don't want them to be able to trial-and-error it).

The next thing that happens is that the Firebase SDK sends this ReCAPTCHA token (good or bad) to the App Check endpoint. Since you've given us your ReCAPTCHA secret key in the Firebase console, the App Check endpoint is able to use that to decode the ReCAPTCHA token and see if it's valid. If it's invalid, it's a 403 error.

At this point, it was decided to throttle any further requests from this client for 24h because it's unlikely that a bad ReCAPTCHA token will resolve itself on an immediate retry, and we didn't want people using up their quota because they keep sending the same bad ReCAPTCHA token to the App Check endpoint. (Other errors, like in the 500 range, where the server was busy or had some error, seem likely to be resolved with a retry in the next few seconds/minutes, so those have a short exponential backoff.)

So there's only two ways I know of to get a bad ReCAPTCHA token (and therefore the 403 which leads to throttling):
1. Wrong site key or secret key other wrong setup - this would lead to an error on every request
2. ReCAPTCHA thinks there is something fishy about your environment - unfortunately this isn't something we can debug easily. ReCAPTCHA is a black box for a reason (mentioned above). You may be able to check your ReCAPTCHA logs on the backend, or you may be able to write a test app that just runs a ReCAPTCHA check end to end (no Firebase) and see if it consistently fails on your mobile or wherever you're seeing the problem.

If the failure rate from ReCAPTCHA fails is unacceptable, one solution is to use `CustomProvider` to write your own attestation provider, instead of using ReCAPTCHA.

## IndexedDB

IndexedDB failures are a separate issue. Firebase App Check Web SDK writes successfully retrieved App Check tokens to the browser's local storage using IndexedDB, and also always checks that same place for an existing token before attempting to retrieve a new one.

We have checks to see if the environment supports IndexedDB, and just avoid using it if the environment does not support it, but it's very tricky because different environments require different checks and it's easy to miss one. If you can narrow this error down to a specific environment, we can quickly make a check for it. It seems like that string has come up in Firefox Private Browsing: https://github.com/ipfs/js-ipfs/issues/1388 but it's not clear from this which operation triggered it, we will have to investigate.


TLDR:

- 403 indicates ReCAPTCHA is failing because, rightly or wrongly, it's finding something suspicious about your environment. Debugging the inner workings of ReCAPTCHA isn't in the scope of this team. Writing your own `CustomProvider` may be a good solution if the failure rate is unacceptable.
- IndexedDB error likely caused by an environment we need to write a new IndexedDB check for. If this error is important, can you provide a minimal repro? This one should fail every time if you're using the specified environment it fails in (probably Firefox Private Browsing?) Otherwise we will try to prioritize looking into it when we can.

**Comment by mpullerits** [comment]
Yes, we are using ReCAPTCHA v3.  How can I check the ReCAPTCHA logs on the backend as you suggested? Is this within the ReCAPTCHA admin somewhere?

What is the difference on ReCAPTCHA between Chrome normal browser, and Chrome private browser? For me, every time I test on my PC, it will fail in private browser, but the exact same URL will work on the same PC on the same Chrome instance, just not private tab. 

This seems very strange to me.

Can you please help me to understand why Chrome normal works, but Chrome private fails?

**Comment by mpullerits** [comment]
I also experience the exact same problem in Firefox private - 

In Firefox normal browser, the page will load totally fine.  The exact same URL, on the exact same PC, yet in Firefox private browser, will fail with the same AppCheck error.

This is very reliable and repeatable. Normal browser, will work, private browser will fail.

What would cause this?

**Comment by DellaBitta** [comment]
Hi @mpullerits,

> Yes, we are using ReCAPTCHA v3. How can I check the ReCAPTCHA logs on the backend as you suggested? Is this within the ReCAPTCHA admin somewhere?

App Check uses ReCAPTCHA as a black box of sorts so we don't have many debug tools on the Firebase SDK side to aid in its administration. I recommend attempting to use [ReCAPTCHA](https://developers.google.com/recaptcha/intro) directly (without AppCheck) in your environments that are causing problems and then see what score it returns to you. If the issue still exists with a low score then you could reach out to  [ReCAPTCHA support](https://developers.google.com/recaptcha/docs/faq) for help. If the problems don't exist, then this is good data for the Firebase Support team, and I recommend opening up a ticket with them. They can work with you to get more data from you and your account privately and would have a better understanding of how the ReCAPTCHA scores are affecting the App Check response.

> Can you please help me to understand why Chrome normal works, but Chrome private fails?

I'm not quite sure. I do know that ReCAPTCHA uses cookies as part of its flow, and this might be the cause of the problem if cookies aren't enabled in your private browsing session? Regardless, I can only speculate. This would be a great question for the ReCAPTCHA support team and/or their stack overflow group.

**Comment by mpullerits** [comment]
Hi DellaBitta,

If we disable AppCheck it will expose us to abuse. How will we be protected from abuse during this testing time as you suggested?

**Comment by ronaldohsn** [comment]
> Can you please help me to understand why Chrome normal works, but Chrome private fails?

In my situation, the issue occurred within the same Chrome instance. It experienced a temporary failure and subsequently returned to a functional state.

While it was functioning correctly on my iPad, the same error "AppCheck: Requests throttled due to 403 error" occurred on another iPad of the same model.

**Comment by hsubox76** [comment]
> Hi DellaBitta,
> 
> If we disable AppCheck it will expose us to abuse. How will we be protected from abuse during this testing time as you suggested?

I think there's a misunderstanding. @DellaBitta was suggesting making a simple test app that only uses ReCAPTCHA, and does something if it passed, and something else if it failed, so you can see which environments it passes or fails in. Not removing App Check from your production app. Once you have more data on this, you can decide what the next steps will be - if ReCAPTCHA works in some environments where App Check fails, let us know which environments and we'll try to reproduce. If ReCAPTCHA fails in every environment where App Check fails for you, and you then decide to request support from the ReCAPTCHA team, you will have more specifics to provide them.
